onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -0,0 +1,232 @@
+# license information.
+# --------------------------------------------------------------------------
+
+import os
+import sys
+import argparse
+import numpy as np
+from PIL import Image
+import onnx
+import onnxruntime
+from onnx import helper, TensorProto, numpy_helper
+from quantize import quantize, QuantizationMode
+from data_preprocess import load_batch
+
+import re
+import subprocess
+import json
+
+def augment_graph(model):
+    '''
+    Adds ReduceMin and ReduceMax nodes to all Conv and MatMul nodes in
+    model and ensures their outputs are stored as part of the graph output
+        parameter model: loaded FP32 ONNX model to quantize
+        return: augmented ONNX model
+    '''
+    added_nodes = []
+    added_outputs = []
+    for node in model.graph.node:
+        if node.op_type in quantization_candidates:
+            input_name = node.output[0]
+            # Adding ReduceMin nodes
+            reduce_min_name = node.name + '_ReduceMin'
+            reduce_min_node = onnx.helper.make_node('ReduceMin', [input_name],
+                            [input_name + '_ReduceMin'], reduce_min_name, keepdims=0)
+            added_nodes.append(reduce_min_node)
+            added_outputs.append(helper.make_tensor_value_info(reduce_min_node.output[0], TensorProto.FLOAT, ()))
+
+            # Adding ReduceMax nodes
+            reduce_max_name = node.name + '_ReduceMax'
+            reduce_max_node = onnx.helper.make_node('ReduceMax', [input_name],
+                            [input_name + '_ReduceMax'], reduce_max_name, keepdims=0)
+            added_nodes.append(reduce_max_node)
+            added_outputs.append(helper.make_tensor_value_info(reduce_max_node.output[0], TensorProto.FLOAT, ()))
+    model.graph.node.extend(added_nodes)
+    model.graph.output.extend(added_outputs)
+    return model
+
+# Using augmented outputs to generate inputs to quantize.py
+def get_intermediate_outputs(model_path, session, inputs, calib_mode='naive'):
+    '''
+    Gather intermediate model outputs after running inference
+        parameter model: path to augmented FP32 ONNX model
+        parameter inputs: list of loaded test inputs (or image matrices)
+        parameter average_mode: type 'naive' gives (ReduceMin, ReduceMax) pairs
+                                for each augmented node across test data sets, where
+                                the first element is a minimum of all ReduceMin values
+                                and the second element is a maximum of all ReduceMax
+                                values; the type 'smooth' yields a smooth average
+                                with anomalous values removed
+        return: dictionary mapping added node names to (ReduceMin, ReduceMax) pairs
+    '''
+    model = onnx.load(model_path)
+    num_model_outputs = len(model.graph.output) # number of outputs in original model
+    num_inputs = len(inputs)
+    input_name = session.get_inputs()[0].name
```

Comment:
This tool only supports one input. Is it by design (for image related model)?

Reply:
calibration data set is a much smaller subset of the training data set so we started with just 1 input. However if need be it can be extended to use multiple inputs in future.